### ISSUE TITLE
Icons are randomly janky (scssphp)

### ISSUE THREAD

**Issue by GameCharmer** [FortAwesome/Font-Awesome#18775]
### closed because it is a duplicate of #17644

---

[](url)I can't quite explain this one, but it happens randomly using the self-hosted SCSS implementation of Font Awesome 6 Pro.  I can go for a few hours without this weird graphic bug occuring.  It's totally random, and I can sit and refresh the same page and watch it crop up once every hundred or so refreshes.  I'm using Chrome on Win 10

Here's some examples that indicate what's going on better than I could put into words.

| ![ex01](https://user-images.githubusercontent.com/1208565/156827520-0534500c-6baf-478d-be75-a6a071539946.png) | ![ex02](https://user-images.githubusercontent.com/1208565/156827521-bfe02ee6-c986-4638-b45f-1426385865ab.png) | ![ex03](https://user-images.githubusercontent.com/1208565/156827523-67d1466f-827e-4d57-8ad4-b6a85d7897c4.png) | ![ex04](https://user-images.githubusercontent.com/1208565/156827524-abfeb6dc-0cc6-4a1e-a6df-649cd019f9fa.png) |
| ![ex11](https://user-images.githubusercontent.com/1208565/156827525-3d160f58-c13f-4d68-8eff-220fc35f9c9d.png) | ![ex12](https://user-images.githubusercontent.com/1208565/156827527-e09ab911-8783-4a27-b8f2-768123759752.png) | ![ex13](https://user-images.githubusercontent.com/1208565/156827528-e18ac1ac-0482-4871-a090-fdbc71c71db5.png) | ![ex14](https://user-images.githubusercontent.com/1208565/156827529-5e8b74ba-b261-449f-94ba-237d69d46efd.png) |

**Comment by tagliala** [comment]
Hi!

Thanks for being part of the Font Awesome Community.

I think this may be related to #17644

Please check how the unicode entities are encoded in the output css file.

If the css contains PUA characters like `content: "";` (instead of `content: "\f26e";`), then I suggest to check that:

1. The HTML file has `<meta charset="utf-8">`
2. The CSS file has `@charset "UTF-8";`

having the charset directive in the HTML only may not be enough

if this is the same issue, let me know so we can continue on #17644

**Comment by GameCharmer** [comment]
@tagliala That was exactly the issue!  I'm not too keen on github issues and didn't realize I was only searching in Open. :P

Sorry about that, have a great day, and thanks for the fast response!

**Comment by GameCharmer** [comment]
@tagliala if I may ask, do you know why it does that at random?  I'm not using the JS, so it should be consistent between page refreshes.

I've tried everything at this point and can't get things to work properly.  If I use the all.css flat out, every character is a broken rectangle on every page.  They do show up as `"\fwhatever` when using the all.css file though.

If I compile the scss, include `@charset "UTF-8";` as the first line, ensure it's being saved as UTF-8 to disk, set the first item in `<head>` to `<meta charset="utf-8">`, the characters show up properly... 99% of the time.  When looking at the compiled.css file, the first line is indeed `@charset "UTF-8";` but all of the content items are `content: "";`

Since we're running NGINX, I even went and added `charset UTF-8;` to the http section in nginx.conf, as well as to every location, still PUA characters in the output, which does work most of the time.

**Comment by tagliala** [comment]
> if I may ask, do you know why it does that at random?

Nope :\

@GameCharmer if I got this correctly, even when both charset directive are set, it does not work 100% of times but it does help, correct?

**Comment by GameCharmer** [comment]
@tagliala correct.  I wonder if this is something specific to chrome.  Reports from end users have dropped off significantly since adding the meta tag and css charset line, and I've only had it crop up once since making the change.  ¯\_(ツ)_/¯

The only question I have left is, should the resulting CSS file `content: "\f26e";` or `content: "";`?

**Comment by tagliala** [comment]
> should the resulting CSS file content: "\f26e";

This. As I mentioned in the other thread, there has been a breaking change in a minor release of dart-sass and ruby-sass

The only workaround I can see is:

- customize the whole `_variables.scss` by removing trailing backslashes from icon variables (eg: `$fa-var-address-book: \f2b9;` => `$fa-var-address-book: f2b9;`)
- override `fa-content` function
  ```scss
  @function fa-content($fa-var){
    @return unquote("\"") + unquote(str-insert($fa-var, "\\", 1)) + unquote("\"");
  }
  ```

Having a copy of icon variables is frustrating but we can't figure out a way to preserve backward compatibility

**Comment by tagliala** [comment]
@GameCharmer 

This should have been fixed a while ago in dart-sass 1.38.0

https://github.com/sass/dart-sass/releases/tag/1.38.0

sass/dart-sass#1430

Please give it a try

**Comment by GameCharmer** [comment]
@tagliala I don't use dart sass at this time.  It looks like there really isn't a good way to handle things in PHP. :P

https://github.com/scssphp/scssphp

**Comment by tagliala** [comment]
Thanks

I should install some dependencies to work with scssphp

Could you please send me the compiled output of 

```scss
$fa-new-content: false !default;

@function fa-content($fa-var) {
  @if $fa-new-content {
    @return unquote("\"") + unquote(str-insert($fa-var, "\\", 1)) + unquote("\"");
  }

  @return unquote("\"#{ $fa-var }\"");
}

$fa-var-500px: \f26e;

@if $fa-new-content {
  $fa-var-500px: "f26e";
}

.icon::before {
  content: fa-content($fa-var-500px);
}
```

with both `$fa-new-content: false` and `$fa-new-content: true`?

**Comment by GameCharmer** [comment]
Output when false
```
@charset "UTF-8";
.icon::before {
  content: "";
}
```

Output when true
```
.icon::before {
  content: "\f26e";
}

```

**Comment by tagliala** [comment]
@GameCharmer if I provide you some instructions to workaround this issue, are you willing to test if that would fix 100% of the cases?

**Comment by GameCharmer** [comment]
@tagliala absolutely

**Comment by tagliala** [comment]
Hi @GameCharmer , I've tried to provide a step by step fix here: diowa/ruby3-rails6-bootstrap-heroku#540

This is based on Font Awesome Free, so steps for Pro should be a little bit different

At the end of the procedure, we should check if variables in the output css have the proper escaped entity

NOTE: This change will make harder to update Font Awesome

Feel free to ask any question

**Comment by GameCharmer** [comment]
@tagliala The CSS generates properly with your fix!  This shows up properly in the CSS, and my pages are rendering properly after numerous refreshes.

```css
.fa-book::before{content:"\f02d"}
```

Thank you!

**Comment by tagliala** [comment]
@GameCharmer welcome

Please let me know if there are changes and you still randomly get failures

**Comment by tagliala** [comment]
I'm going to lock here and move the discussion to #17644

**Comment by tagliala** [comment]
Reopening because of https://github.com/FortAwesome/Font-Awesome/issues/17644#issuecomment-1259578128

**Comment by tagliala** [comment]
@GameCharmer unlocked, please report here scssphp specific issues, I will then move to the other thread if they are relevant

**Comment by GameCharmer** [comment]
@tagliala It looks like I got something wrong between fontawesome versions when rewriting the config file.  Previously, I had something like this

```php
$file    = $path.'_variables.scss';
$content = \file_get_contents($file);
$data    = explode(PHP_EOL, $content);

foreach($data as $i => $line) {
    if (!str_contains($line, '\\') || !str_contains($line,';')) {
        continue;
    }
    $line = str_replace(';','";',$line);
    $line = str_replace('\\','"',$line);

    $data[$i] = $line;
}

\file_put_contents($file, implode(PHP_EOL, $data);)
```
Upon changing `$line = str_replace('\\','"',$line);` to `$line = str_replace('\\','"\\',$line);` for 6.2, we're good to go again.

**Comment by GameCharmer** [comment]
Ahh crap, dev and production didn't line up, jumped the gun on the response.

When applying the same set of fixes as last time, I wind up with resulting classes like so

```css
.fa-chart-network::before{content:"f78a"}
```

but the rendered output results in this
![image](https://user-images.githubusercontent.com/1208565/192681882-bcb334a1-67a8-489b-b2d6-de5bfb604a67.png)



I've swapped out the fa-content function in _functions.scss to the below
```scss
@function fa-content($fa-var) {
  @return unquote("\"") + unquote(str-insert($fa-var, "\\", 1)) + unquote("\"");
}
```
 and have updated _variables.scss 
 ```scss
 ...
 $fa-var-fill-drip: "f576";
$fa-var-arrows-to-circle: "e4bd";
$fa-var-circle-chevron-right: "f138";
$fa-var-chevron-circle-right: "f138";
$fa-var-wagon-covered: "f8ee";
$fa-var-line-height: "f871";
$fa-var-bagel: "e3d7";
$fa-var-transporter-7: "e2a8";
$fa-var-at: "40";
...
```

**Comment by GameCharmer** [comment]
Interesting enough, this time around, if I avoid scss entirely and just load in `all.css` and run a simple str_replace, I'm golden.

```php
    $faCss = \file_get_contents($faDir.'css/all.css');
    $faCss = str_replace('../webfonts/',$faWeb.'webfonts/',$faCss);
```

That works perfect.  Sometimes the simplest option is the best option.

**Comment by magynhard** [comment]
For others, facing the same issue:

The problem with displaying the icons incorrectly when contained in unicode characters is Chrome related and happens only in Chrome and other Chromium based browsers (e.g. Edge or Vivaldi). I could not reproduce it on Firefox or Safari.

It is not related to FontAwesome, it is related to all Iconfonts using unicode. (Im my case my own font)

It can even be reproduced regularly the following way:
* Open your browser
* Open your website
* Open Chrome Dev Tools (F12)
* Reload Page (only F5 or Reload button, not Force Reload)
* Icons are displayed incorrectly

It works only once, if you want to repeat, you need to close the developer tools, reload the page, then reopen the developer tools again and reload again.

It seems, that chrome does not check or detect the format of the CSS correctly when the page is reloaded with developer tools open at the first time.

For me, the best solution was - as mentioned by others before - to add
```
@charset "UTF-8";
```
at the beginning(!) of the .css file to workaround the Chromium engine bug.
